src/coreclr/jit/lower.cpp

```diff
@@ -4317,8 +4317,14 @@ GenTree* Lowering::LowerJTrue(GenTreeOp* jtrue)
         if (cond->OperIs(GT_EQ, GT_NE) && relopOp2->IsIntegralConst(0))
         {
             // Codegen will use cbz or cbnz in codegen which do not affect the flag register
-            newOper = GT_JCMP;
-            cc      = GenCondition::FromRelop(cond);
+            // Exception to this transformation is if relopOp1 can set flags and we expect TryLowerConditionToFlagsNode will later 
+            // transform the tree to reuse the zero flag, then we can avoid changing to compare + branch here.
+            // TryLowerConditionToFlagsNode will only transform if optimizations enabled so don't bother avoiding this change if optimizations is off
+            if (!comp->opts.OptimizationEnabled() || !relopOp1->SupportsSettingZeroFlag() || !IsInvariantInRange(relopOp1, jtrue))
+            {
+                newOper = GT_JCMP;
+                cc      = GenCondition::FromRelop(cond);
+            }
```

Comment:
Can you give an example of the cases this improves? Generally I would not expect this to be helpful -- IIUC it will just result in changing something like `and x, 123; cbz x <target>` to `ands x, 123; beq <target>`.

Reply:
yes, it is generally only and/bic, cbz -> ands/bics, beq

I am expecting that doing a compare after a result that already sets the zero flag is extra work, so ands/bics + branch should be faster. Do I have a mistaken understanding?

Reply:
The way to figure that out is to measure it on some hardware. It's hard to say what the impact of something like that may be. I would expect both patterns to be handled equally efficiently by modern ARM64 hardware.

Reply:
Good idea I will do that, thank you.

I have a couple of other questions:
1. on x64, I tried to turn `and + cmp` into `not + test` instead of `andn`, but I found that the data (already in a register) that will be `not`'d will be copied to a scratch register first. I think this defeats the purpose of the `not + test` structure. If the register can be used, it would be better to keep `andn`.

2. on arm64, it is also similarly more efficient (in my mind) that if the constant value can be used in an `and` instruction, it would be better to keep `and + cmp` as opposed to `bics` because `bics` requires register arguments, so getting the constant into a register requires a `mov`. 

For these both, would your recommendation to also test it on hardware? I don't have much variety when it comes to testing hardware, so not sure if I can extrapolate that data to all cases.